src/net_processing.cpp

```diff
@@ -3720,15 +3754,15 @@ void ProcessMessage(
 bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
 {
     const NodeId peer_id{pnode.GetId()};
+    PeerRef peer = GetPeerRef(peer_id);
```

Comment:
36ac41a

- perhaps this line inside the block like the three cases above

- should we add an `if (peer == nullptr) return false;` check here as well?

Reply:
A reference to the node exists (`CNode& pnode`) so it should be impossible for the peer ref to be deleted already

Reply:
> perhaps this line inside the block like the three cases above

I don't think this is required. The block is to limit the scope of the lock. Putting the `peer` variable in there is fine, but I don't think it's consistent with project style.

The block in `GetNodeStateStats()` was left over from a previous iteration of this branch. I've removed it. The block in `FinalizeNode()` is to force the `PeerRef` object to go out of scope, which will generally cause the underlying `Peer` object to be destructed, since that's usually the last reference.

> should we add an if (peer == nullptr) return false; check here as well?

Yes we should. Thank you!